File: python/my_first_package/my_first_package/services/my_first_service/handler.py
from . import dynamic as my_first_package_services_my_first_service_dynamic
from . import structures as my_first_package_services_my_first_service_structures
from typelayer_common import application as typelayer_common_application
from typelayer_common import documentation as typelayer_common_documentation
from typelayer_common import dynamic as typelayer_common_dynamic
from dataclasses import dataclass
from typing import List, Dict, Optional


class MyFirstService:
    def my_first_service_my_first_function(self, incoming_msg: my_first_package_services_my_first_service_structures.MyFirstFunctionInput) -> my_first_package_services_my_first_service_structures.MyFirstFunctionOutput:
        raise NotImplementedError
    def my_first_service_my_second_function(self, incoming_msg: my_first_package_services_my_first_service_structures.MySecondFunctionInput) -> my_first_package_services_my_first_service_structures.MySecondFunctionOutput:
        raise NotImplementedError

@dataclass
class MyFirstServiceApp(typelayer_common_application.Process):
    service: MyFirstService;
    def process_describe(self: "MyFirstServiceApp") -> typelayer_common_documentation.ApplicationDescription:
    
        service_description = describe_my_first_service()
        return typelayer_common_documentation.ApplicationDescription(name="my_first_service", description=None, services=[service_description])
    def process_handle(self: "MyFirstServiceApp", encoding: typelayer_common_dynamic.Encoding, path: List[str], body: bytes) -> bytes:
        
        
        if (path[0] == "my-first-function"):
            response = handle_my_first_service_my_first_function(self.service, encoding, body)
        elif (path[0] == "my-second-function"):
            response = handle_my_first_service_my_second_function(self.service, encoding, body)
        else:
                raise typelayer_common_application.FunctionNotFound("Function not found", {"requested_service": "my-first-service", "requested_function": path[0], "available_functions": "my-first-function | my-second-function"});
        self.service = self.service;
        
        resp_bin = response.encoder_to_binary()
        return resp_bin

def describe_my_first_service() -> typelayer_common_documentation.ServiceDescription:
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    my_first_argument_field_info = typelayer_common_documentation.TlText()
    my_first_argument_info = typelayer_common_documentation.FieldDescription(name="my_first_argument", description=None, type_info=my_first_argument_field_info)
    my_first_function_inputs = [my_first_argument_info]
    my_first_return_field_info = typelayer_common_documentation.TlStructure(package="my_first_package", schema="structures", structure_id="my_first_structure")
    my_first_return_info = typelayer_common_documentation.FieldDescription(name="my_first_return", description=None, type_info=my_first_return_field_info)
    my_first_function_outputs = [my_first_return_info]
    arg_one_field_info = typelayer_common_documentation.TlStructure(package="my_first_package", schema="structures", structure_id="my_enum_structure")
    arg_one_info = typelayer_common_documentation.FieldDescription(name="arg_one", description=None, type_info=arg_one_field_info)
    arg_two_field_info = typelayer_common_documentation.TlText()
    arg_two_info = typelayer_common_documentation.FieldDescription(name="arg_two", description=None, type_info=arg_two_field_info)
    my_second_function_inputs = [arg_one_info, arg_two_info]
    ret_one_inner_field_info = typelayer_common_documentation.TlText()
    ret_one_field_info = typelayer_common_documentation.TlList(list_inner=ret_one_inner_field_info)
    ret_one_info = typelayer_common_documentation.FieldDescription(name="ret_one", description=None, type_info=ret_one_field_info)
    my_second_function_outputs = [ret_one_info]
    return typelayer_common_documentation.ServiceDescription(name="my_first_service", description=None, functions=[typelayer_common_documentation.FunctionDescription(name="my_first_service", description=None, inputs=my_first_function_inputs, outputs=my_first_function_outputs), typelayer_common_documentation.FunctionDescription(name="my_first_service", description=None, inputs=my_second_function_inputs, outputs=my_second_function_outputs)])

def handle_my_first_service_my_first_function(implementation: MyFirstService, encoding: typelayer_common_dynamic.Encoding, binary: bytes) -> typelayer_common_dynamic.Encoder:
    
    
    
    
    decoder_start = encoding.encoding_start_decoding(binary)
    encoder_start = encoding.encoding_start_encoding()
    input = my_first_package_services_my_first_service_dynamic.dynamic_get_obj_my_first_function_input(decoder_start)
    result = implementation.my_first_service_my_first_function(input)
    my_first_package_services_my_first_service_dynamic.dynamic_put_my_first_function_output(encoder_start, result)
    return encoder_start

def handle_my_first_service_my_second_function(implementation: MyFirstService, encoding: typelayer_common_dynamic.Encoding, binary: bytes) -> typelayer_common_dynamic.Encoder:
    
    
    
    
    decoder_start = encoding.encoding_start_decoding(binary)
    encoder_start = encoding.encoding_start_encoding()
    input = my_first_package_services_my_first_service_dynamic.dynamic_get_obj_my_second_function_input(decoder_start)
    result = implementation.my_first_service_my_second_function(input)
    my_first_package_services_my_first_service_dynamic.dynamic_put_my_second_function_output(encoder_start, result)
    return encoder_start

File: typescript/my-first-package/services/myFirstService/handler.ts
// Module: services.myFirstService.handler
import * as myFirstPackageServicesMyFirstServiceDynamic from "./dynamic";
import * as myFirstPackageServicesMyFirstServiceStructures from "./structures";
import * as typelayerCommonApplication from "typelayer-common";
import * as typelayerCommonDocumentation from "typelayer-common";
import * as typelayerCommonDynamic from "typelayer-common";


export interface MyFirstService {
    myFirstServiceMyFirstFunction: (incomingMsg: myFirstPackageServicesMyFirstServiceStructures.MyFirstFunctionInput) => Promise<myFirstPackageServicesMyFirstServiceStructures.MyFirstFunctionOutput>;
    myFirstServiceMySecondFunction: (incomingMsg: myFirstPackageServicesMyFirstServiceStructures.MySecondFunctionInput) => Promise<myFirstPackageServicesMyFirstServiceStructures.MySecondFunctionOutput>;
}

export class MyFirstServiceApp implements typelayerCommonApplication.Process {
    kind : string = "MyFirstServiceApp";
    service: MyFirstService;
    constructor(arg: {service: MyFirstService;}) {
        Object.assign(this, arg);
    }
    processDescribe(this: MyFirstServiceApp): typelayerCommonDocumentation.ApplicationDescription {
        let serviceDescription : typelayerCommonDocumentation.ServiceDescription;
        serviceDescription = describeMyFirstService();
        return new typelayerCommonDocumentation.ApplicationDescription({name: "my_first_service", description: null, services: [serviceDescription]});
    }
    async processHandle(this: MyFirstServiceApp, encoding: typelayerCommonDynamic.Encoding, path: Array<string>, body: Uint8Array): Promise<Uint8Array> {
        let respBin : Uint8Array;
        let response : typelayerCommonDynamic.Encoder;
        switch (path[0]) {
            case "my-first-function":
                response = await handleMyFirstServiceMyFirstFunction(this.service, encoding, body);
                break;
            case "my-second-function":
                response = await handleMyFirstServiceMySecondFunction(this.service, encoding, body);
                break;
            default:
                throw new typelayerCommonApplication.FunctionNotFound("Function not found", {requestedService: "my-first-service", requestedFunction: path[0], availableFunctions: "my-first-function | my-second-function"});
        }
        this.service = this.service;
        
        respBin = response.encoderToBinary();
        return respBin;
    }
}

export function describeMyFirstService(): typelayerCommonDocumentation.ServiceDescription {
    let argOneFieldInfo : typelayerCommonDocumentation.TypeInfo;
    let argOneInfo : typelayerCommonDocumentation.FieldDescription;
    let argTwoFieldInfo : typelayerCommonDocumentation.TypeInfo;
    let argTwoInfo : typelayerCommonDocumentation.FieldDescription;
    let myFirstArgumentFieldInfo : typelayerCommonDocumentation.TypeInfo;
    let myFirstArgumentInfo : typelayerCommonDocumentation.FieldDescription;
    let myFirstFunctionInputs : Array<typelayerCommonDocumentation.FieldDescription>;
    let myFirstFunctionOutputs : Array<typelayerCommonDocumentation.FieldDescription>;
    let myFirstReturnFieldInfo : typelayerCommonDocumentation.TypeInfo;
    let myFirstReturnInfo : typelayerCommonDocumentation.FieldDescription;
    let mySecondFunctionInputs : Array<typelayerCommonDocumentation.FieldDescription>;
    let mySecondFunctionOutputs : Array<typelayerCommonDocumentation.FieldDescription>;
    let retOneFieldInfo : typelayerCommonDocumentation.TypeInfo;
    let retOneInfo : typelayerCommonDocumentation.FieldDescription;
    let retOneInnerFieldInfo : typelayerCommonDocumentation.TypeInfo;
    myFirstArgumentFieldInfo = new typelayerCommonDocumentation.TlText({});
    myFirstArgumentInfo = new typelayerCommonDocumentation.FieldDescription({name: "my_first_argument", description: null, typeInfo: myFirstArgumentFieldInfo});
    myFirstFunctionInputs = [myFirstArgumentInfo];
    myFirstReturnFieldInfo = new typelayerCommonDocumentation.TlStructure({package: "my_first_package", schema: "structures", structureId: "my_first_structure"});
    myFirstReturnInfo = new typelayerCommonDocumentation.FieldDescription({name: "my_first_return", description: null, typeInfo: myFirstReturnFieldInfo});
    myFirstFunctionOutputs = [myFirstReturnInfo];
    argOneFieldInfo = new typelayerCommonDocumentation.TlStructure({package: "my_first_package", schema: "structures", structureId: "my_enum_structure"});
    argOneInfo = new typelayerCommonDocumentation.FieldDescription({name: "arg_one", description: null, typeInfo: argOneFieldInfo});
    argTwoFieldInfo = new typelayerCommonDocumentation.TlText({});
    argTwoInfo = new typelayerCommonDocumentation.FieldDescription({name: "arg_two", description: null, typeInfo: argTwoFieldInfo});
    mySecondFunctionInputs = [argOneInfo, argTwoInfo];
    retOneInnerFieldInfo = new typelayerCommonDocumentation.TlText({});
    retOneFieldInfo = new typelayerCommonDocumentation.TlList({listInner: retOneInnerFieldInfo});
    retOneInfo = new typelayerCommonDocumentation.FieldDescription({name: "ret_one", description: null, typeInfo: retOneFieldInfo});
    mySecondFunctionOutputs = [retOneInfo];
    return new typelayerCommonDocumentation.ServiceDescription({name: "my_first_service", description: null, functions: [new typelayerCommonDocumentation.FunctionDescription({name: "my_first_service", description: null, inputs: myFirstFunctionInputs, outputs: myFirstFunctionOutputs}), new typelayerCommonDocumentation.FunctionDescription({name: "my_first_service", description: null, inputs: mySecondFunctionInputs, outputs: mySecondFunctionOutputs})]});
}

export async function handleMyFirstServiceMyFirstFunction(implementation: MyFirstService, encoding: typelayerCommonDynamic.Encoding, binary: Uint8Array): Promise<typelayerCommonDynamic.Encoder> {
    let decoderStart : typelayerCommonDynamic.Decoder;
    let encoderStart : typelayerCommonDynamic.Encoder;
    let input : myFirstPackageServicesMyFirstServiceStructures.MyFirstFunctionInput;
    let result : myFirstPackageServicesMyFirstServiceStructures.MyFirstFunctionOutput;
    decoderStart = encoding.encodingStartDecoding(binary);
    encoderStart = encoding.encodingStartEncoding();
    input = myFirstPackageServicesMyFirstServiceDynamic.dynamicGetObjMyFirstFunctionInput(decoderStart);
    result = await implementation.myFirstServiceMyFirstFunction(input);
    myFirstPackageServicesMyFirstServiceDynamic.dynamicPutMyFirstFunctionOutput(encoderStart, result);
    return encoderStart;
}

export async function handleMyFirstServiceMySecondFunction(implementation: MyFirstService, encoding: typelayerCommonDynamic.Encoding, binary: Uint8Array): Promise<typelayerCommonDynamic.Encoder> {
    let decoderStart : typelayerCommonDynamic.Decoder;
    let encoderStart : typelayerCommonDynamic.Encoder;
    let input : myFirstPackageServicesMyFirstServiceStructures.MySecondFunctionInput;
    let result : myFirstPackageServicesMyFirstServiceStructures.MySecondFunctionOutput;
    decoderStart = encoding.encodingStartDecoding(binary);
    encoderStart = encoding.encodingStartEncoding();
    input = myFirstPackageServicesMyFirstServiceDynamic.dynamicGetObjMySecondFunctionInput(decoderStart);
    result = await implementation.myFirstServiceMySecondFunction(input);
    myFirstPackageServicesMyFirstServiceDynamic.dynamicPutMySecondFunctionOutput(encoderStart, result);
    return encoderStart;
}